src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/MethodToStateMachineRewriter.cs

```diff
@@ -465,11 +465,12 @@ private BoundExpression HoistRefInitialization(SynthesizedLocal local, BoundAssi
             bool needsSacrificialEvaluation = false;
             var hoistedFields = ArrayBuilder<StateMachineFieldSymbol>.GetInstance();
 
-            AwaitExpressionSyntax awaitSyntaxOpt;
+            SyntaxNode awaitSyntaxOpt;
             int syntaxOffset;
             if (F.Compilation.Options.OptimizationLevel == OptimizationLevel.Debug)
             {
-                awaitSyntaxOpt = (AwaitExpressionSyntax)local.GetDeclaratorSyntax();
+                awaitSyntaxOpt = local.GetDeclaratorSyntax();
+                Debug.Assert(awaitSyntaxOpt.IsKind(SyntaxKind.AwaitExpression) || awaitSyntaxOpt.IsKind(SyntaxKind.SwitchExpression));
```

Comment:
Could you help me understand why only `await` and `switch` are expected to be encountered here? For example if I had `GetRef() = flag ? await GetAsync() : 42` would I hit this assert?

Reply:
You would not hit this assert. There's some differences in the syntaxes used by the SpillSequenceSpiller here, and the local in the case you mentioned would be the await expression, not the ternary. I'd need to dig more into the local rewriter to tell you the exact reasons for that difference.